src/stdio_mgr/types.py

```diff
@@ -0,0 +1,179 @@
+r"""``stdio_mgr.types`` *code module*.
+
+``stdio_mgr.types`` provides context managers for convenient
+interaction with ``stdin``/``stdout``/``stderr`` as a tuple.
+
+**Author**
+    John Vandenberg (jayvdb@gmail.com)
+
+**File Created**
+    6 Sep 2019
+
+**Copyright**
+    \(c) Brian Skinn 2018-2019
+
+**Source Repository**
+    http://www.github.com/bskinn/stdio-mgr
+
+**Documentation**
+    See README.rst at the GitHub repository
+
+**License**
+    The MIT License; see |license_txt|_ for full license terms
+
+**Members**
+
+"""
+import abc
+import sys
+from contextlib import ExitStack, suppress
+from io import TextIOBase
+
+import _collections_abc
+
+# AbstractContextManager was introduced in Python 3.6
+# and may be used with typing.ContextManager.
+try:
+    from contextlib import AbstractContextManager
+except ImportError:  # pragma: no cover
+
+    class AbstractContextManager(abc.ABC):
+        """An abstract base class for context managers."""
+
+        def __enter__(self):
+            """Return `self` upon entering the runtime context."""
+            return self
+
+        @abc.abstractmethod
+        def __exit__(self, exc_type, exc_value, traceback):
+            """Raise any exception triggered within the runtime context."""
+            return None
+
+        @classmethod
+        def __subclasshook__(cls, subclass):
+            """Check whether subclass is considered a subclass of this ABC."""
+            if cls is AbstractContextManager:
+                return _collections_abc._check_methods(
+                    subclass, "__enter__", "__exit__"
+                )
+            return NotImplemented
+
+
+class TupleContextManager(tuple, AbstractContextManager):
+    """Base for context managers that are also a tuple."""
+
+    # This is needed to establish a workable MRO.
+
+
+class StdioTuple(TupleContextManager):
```

Comment:
In the case where fewer than the three full streams are mocked/wrapped by stdio-mgr, are you envisioning still carrying references to all three in, e.g., `StdioTuple`? Just, the ones that aren't mocked/wrapped are the original objects?

Reply:
Yes.  I think that is best, as it means the algorithms are always the same wrt member offsets, and the idea I have in my head is that we would have meta-algorithms about transforms/transitions from one tuple to another being active, e.g. a stack of them, and also sanity checks.

Reply:
&lt;nod&gt;, I agree that this is probably best...push all of the complexity of coping with different states/statuses/identities of the various streams into one place, the logic for in(tro)specting the composition of the 3-tuples.

And, yeah--having different-length tuples seems like a recipe for infinite headache.